FirebaseDatabaseUI/FirebaseIndexArray.m

```diff
@@ -0,0 +1,205 @@
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+// clang-format on
+
+#import "FirebaseIndexArray.h"
+#import "FirebaseQueryObserver.h"
+
+@interface FirebaseIndexArray () <FirebaseArrayDelegate>
+
+@property (nonatomic, readonly) id<FIRDataObservable> index;
+@property (nonatomic, readonly) id<FIRDataObservable> data;
+
+@property (nonatomic, readonly) FirebaseArray *indexArray;
+
+@property (nonatomic, readonly) NSMutableArray<FirebaseQueryObserver *> *observers;
+
+@end
+
+@implementation FirebaseIndexArray
+
+- (instancetype)init {
+  NSException *e =
+    [NSException exceptionWithName:@"FIRUnavailableMethodException"
+                            reason:@"-init is unavailable. Please use the designated initializer instead."
+                          userInfo:nil];
+  @throw e;
+}
+
+- (instancetype)initWithIndex:(id<FIRDataObservable>)index
+                         data:(id<FIRDataObservable>)data
+                     delegate:(nullable id<FirebaseIndexArrayDelegate>)delegate; {
+  NSParameterAssert(index != nil);
+  NSParameterAssert(data != nil);
+  self = [super init];
+  if (self != nil) {
+    _index = index;
+    _data = data;
+    _observers = [NSMutableArray array];
+    _delegate = delegate;
+    [self observeQueries];
+  }
+  return self;
+}
+
+- (instancetype)initWithIndex:(id<FIRDataObservable>)index
+                         data:(id<FIRDataObservable>)data {
+  return [self initWithIndex:index data:data delegate:nil];
+}
+
+- (void)observeQueries {
+  _indexArray = [[FirebaseArray alloc] initWithQuery:self.index delegate:self];
+}
+
+- (NSArray <FIRDataSnapshot *> *)items {
+  NSArray *observers = [self.observers copy];
+  NSMutableArray *array = [[NSMutableArray alloc] initWithCapacity:observers.count];
+  for (FirebaseQueryObserver *observer in observers) {
+    if (observer.contents != nil) {
+      [array addObject:observer.contents];
+    }
+  }
+  return [array copy];
+}
+
+- (NSUInteger)count {
+  return self.observers.count;
+}
+
+// FirebaseIndexArray instance becomes unusable after invalidation.
+- (void)invalidate {
+  for (NSInteger i = 0; i < self.observers.count; i++) {
+    FirebaseQueryObserver *observer = self.observers[i];
+    [observer removeAllObservers];
+  }
+  _observers = nil;
+}
+
+- (FIRDataSnapshot *)objectAtIndex:(NSUInteger)index {
+  return self.observers[index].contents;
+}
+
+- (void)dealloc {
+  [self invalidate];
+}
+
+#pragma mark - FirebaseArrayDelegate
+
+// These delegate methods are entirely responsible
+// for keeping a local array of queries up to date.
```

Comment:
This comment is a bit abandoned here. I might be inclined to move it into a top level class comment that explains what we are trying to do and how we go about it.